pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/merge_rollup/MergeRollupTaskGenerator.java

```diff
@@ -0,0 +1,491 @@
+        for (PinotTaskConfig taskConfig : _clusterInfoAccessor.getTaskConfigs(entry.getKey())) {
+          inCompleteMergeLevels.add(taskConfig.getConfigs().get(MergeRollupTask.MERGE_LEVEL_KEY));
+        }
+      }
+
+      ZNRecord mergeRollupTaskZNRecord = _clusterInfoAccessor.getMinionMergeRollupTaskZNRecord(offlineTableName);
+      int expectedVersion = mergeRollupTaskZNRecord != null ? mergeRollupTaskZNRecord.getVersion() : -1;
+      MergeRollupTaskMetadata mergeRollupTaskMetadata =
+          mergeRollupTaskZNRecord != null ? MergeRollupTaskMetadata.fromZNRecord(mergeRollupTaskZNRecord)
+              : new MergeRollupTaskMetadata(offlineTableName, new TreeMap<>());
+      List<PinotTaskConfig> pinotTaskConfigsForTable = new ArrayList<>();
+
+      // Schedule tasks from lowest to highest merge level (e.g. Hourly -> Daily -> Monthly -> Yearly)
+      for (int i = 0; i < sortedMergeLevelConfigs.size(); i++) {
+        String lowerMergeLevel = i > 0 ? sortedMergeLevelConfigs.get(i - 1).getKey() : null;
+        String mergeLevel = sortedMergeLevelConfigs.get(i).getKey();
+        Map<String, String> mergeConfigs = sortedMergeLevelConfigs.get(i).getValue();
+
+        // Skip scheduling if there's incomplete task for current mergeLevel
+        if (inCompleteMergeLevels.contains(mergeLevel)) {
+          LOGGER.info("Found incomplete task of merge level: {} for the same table: {}, Skipping task generation: {}",
+              mergeLevel, offlineTableName, taskType);
+          continue;
+        }
+
+        // Get the bucket size and buffer
+        long bucketMs =
+            TimeUtils.convertPeriodToMillis(mergeConfigs.get(MinionConstants.MergeTask.BUCKET_TIME_PERIOD_KEY));
+        long bufferMs =
+            TimeUtils.convertPeriodToMillis(mergeConfigs.get(MinionConstants.MergeTask.BUFFER_TIME_PERIOD_KEY));
+
+        // Get watermark from MergeRollupTaskMetadata ZNode
+        // windowStartMs = watermarkMs, windowEndMs = windowStartMs + bucketTimeMs
+        long waterMarkMs =
+            getWatermarkMs(preSelectedSegments.first().getStartTimeMs(), bucketMs, mergeLevel, mergeRollupTaskMetadata);
+        long windowStartMs = waterMarkMs;
+        long windowEndMs = windowStartMs + bucketMs;
+
+        if (!isValidMergeWindowEndTime(windowEndMs, bufferMs, lowerMergeLevel, mergeRollupTaskMetadata)) {
+          LOGGER.info(
+              "Window with start: {} and end: {} of mergeLevel: {} is not a valid merge window, Skipping task generation: {}",
+              windowStartMs, windowEndMs, mergeLevel, taskType);
+          continue;
+        }
+
+        // Find all segments overlapping with the merge window, if all overlapping segments are merged, bump up the target window
+        List<OfflineSegmentZKMetadata> selectedSegments = new ArrayList<>();
+        boolean hasUnMergedSegments = false;
+        boolean isValidMergeWindow = true;
+
+        // The for loop terminates in following cases:
```

Comment:
There are too many conditions which might cause several corner cases. I think we can simplify the logic to only bump up the watermark when unmerged segments are found (new task to be scheduled). This should be able to handle all the corner cases and is much simpler

Reply:
Good point! Updated to bump up watermark only if we need to schedule.